test/common/upstream/cluster_manager_impl_test.cc

```diff
@@ -164,6 +167,167 @@ envoy::config::bootstrap::v3::Bootstrap defaultConfig() {
+}
+
+TEST_F(ODCDTest, TestDiscoveryManagerIgnoresIrrelevantClusters) {
+  auto cb = std::make_shared<ClusterDiscoveryCallback>([](ClusterDiscoveryStatus) {
+    ADD_FAILURE() << "The callback should not be called for irrelevant clusters";
+  });
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo"));
+  auto handle = odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", cb, timeout_);
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_irrelevant"), "version1");
+}
+
+TEST_F(ODCDTest, TestDroppingHandles) {
+  auto cb1 = std::make_shared<ClusterDiscoveryCallback>(
+      [](ClusterDiscoveryStatus) { ADD_FAILURE() << "The callback 1 should not be called"; });
+  auto cb2 = std::make_shared<ClusterDiscoveryCallback>(
+      [](ClusterDiscoveryStatus) { ADD_FAILURE() << "The callback 2 should not be called"; });
+  auto cb3 = std::make_shared<ClusterDiscoveryCallback>(
+      [](ClusterDiscoveryStatus) { ADD_FAILURE() << "The callback 3 should not be called"; });
+  auto cb4 = std::make_shared<ClusterDiscoveryCallback>(
+      [](ClusterDiscoveryStatus) { ADD_FAILURE() << "The callback 4 should not be called"; });
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo1"));
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo2"));
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo3"));
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo4"));
+  auto handle1 = odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo1", cb1, timeout_);
+  auto handle2 = odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo2", cb2, timeout_);
+  auto handle3 = odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo3", cb3, timeout_);
+  auto handle4 = odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo4", cb4, timeout_);
+
+  handle2.reset();
+  handle3.reset();
+  handle1.reset();
+  handle4.reset();
+
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo1"), "version1");
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo2"), "version1");
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo3"), "version1");
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo4"), "version1");
+}
+
+TEST_F(ODCDTest, TestCallbackWithExistingCluster) {
+  auto cb = createCallback(ClusterDiscoveryStatus::Available);
+  EXPECT_CALL(*this, testFunction());
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo"), "version1");
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo")).Times(0);
+  auto handle = odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", cb, timeout_);
+}
+
+TEST_F(ODCDTest, TestExpiredCallbackWithExistingCluster) {
+  cluster_manager_->addOrUpdateCluster(defaultStaticCluster("cluster_foo"), "version1");
+  ClusterDiscoveryCallbackSharedPtr null_cb;
+  EXPECT_CALL(*odcds_, updateOnDemand("cluster_foo")).Times(0);
+  auto handle = odcds_handle_->requestOnDemandClusterDiscovery("cluster_foo", null_cb, timeout_);
+}
```

Comment:
I think `requestOnDemandClusterDiscovery` needs more tests, considering how central it is and the amount of state it changes. I don't think any of these are checked here: timer creation, updates to pending_cluster_creations_, callback handles being returned and if they are the ones that are expected, handling of clusters that were discovered after disovery request has been posted on the main thread.

Reply:
It's difficult to test this function in depth with the current state of mocks (like mock dispatcher executing callbacks immediately). I'll have a look what can be added.

Reply:
> timer creation,

This happens when ODCDS is being invoked, and we test this scenario. Not sure what else to do here.

> updates to pending_cluster_creations_,

These happen when about to invoke ODCDS, when adding the cluster or when discovery expires. I believe the tests cover those three cases.

> callback handles being returned and if they are the ones that are expected,

I'll add a test for it.

> handling of clusters that were discovered after disovery request has been posted on the main thread.

I don't think it's possible to test that with the current mock dispatcher, because it behaves like there is only one thread.